### PR TITLE
Fix multi_user_dm test

### PR DESCRIPTION
**Pull request by asdil12**
Ticket: https://progress.opensuse.org/issues/29986
Testrun: http://artemis.suse.de/tests/193

- Reboot to restart X11 to workaround graphics problems
- Login as test1 and not as bernhard

### CODE REVIEW

File: tests/x11/multi_users_dm.pm [hunk 1]
```diff
@@ -38,11 +38,22 @@ sub ensure_graphical_target {
 }
 
 sub restart_x11 {
-    ensure_multi_user_target;
-    ensure_graphical_target;
+    my ($self, $setnologin) = @_;
+    record_info('workaround', 'Graphical errors on cirrus when restarting X11 -> performing reboot (see bsc#1063060)');
```
Comment (rwx788) [inline tests/x11/multi_users_dm.pm]: Let's write soft_failure here instead, as it's a bug.

File: tests/x11/multi_users_dm.pm [hunk 2]
```diff
@@ -73,7 +84,7 @@ sub run {
         type_string "$user\n";
         send_key 'tab';
     }
-    handle_login;
+    handle_login('user1', 1);
```
Comment (rwx788) [inline tests/x11/multi_users_dm.pm]: I guess we want to use $user variable here.

File: tests/x11/multi_users_dm.pm [hunk 3]
```diff
@@ -89,7 +100,7 @@ sub run {
     script_run "mv /etc/sysconfig/displaymanager.bak /etc/sysconfig/displaymanager";
     assert_script_run "~$username/data/delete_users $users_to_create";
     script_run "clear";
-    restart_x11;
+    $self->restart_x11(0);
```
Comment (rwx788) [inline tests/x11/multi_users_dm.pm]: We don't really need to pass 0, as $setnologin will be undef and both are false. But that's up to you.

File: tests/x11/multi_users_dm.pm [hunk 4]
```diff
@@ -38,11 +38,22 @@ sub ensure_graphical_target {
...
-    ensure_graphical_target;
+    my ($self, $setnologin) = @_;
+    record_info('workaround', 'Graphical errors on cirrus when restarting X11 -> performing reboot (see bsc#1063060)');
+    type_string "reboot\n";
+    my $timeout = get_var('UEFI') ? 140 : 80;
+    if ($setnologin) {
+        $self->wait_boot(bootloader_time => $timeout, nologin => 1, reallynologin => 1);
+    }
+    else {
+        $self->wait_boot(bootloader_time => $timeout);
```
Comment (rwx788) [inline tests/x11/multi_users_dm.pm]: We can combine these two branches in the single call by setting params to $setnologin
```perl
nologin => $setnologin, reallynologin => $setnologin
```

**Comment by asdil12** [comment]
Verification run: http://artemis.suse.de/tests/200

File: lib/utils.pm [hunk 5]
```diff
@@ -841,6 +841,10 @@ sub random_string {
 }
 
 sub handle_login {
+    my ($myuser, $user_selected) = @_;
+    $myuser        = defined $myuser        ? $myuser        : $username;
+    $user_selected = defined $user_selected ? $user_selected : 0;
```
Comment (okurz) [inline lib/utils.pm]: I think you want

```perl
$myuser //= $username;
$user_selected //= 0;
```

File: lib/utils.pm [hunk 6]
```diff
@@ -841,6 +841,10 @@ sub random_string {
 }
 
 sub handle_login {
```
Comment (okurz) [inline lib/utils.pm]: Please put a POD doc above in format as e.g. for `assert_screen_with_soft_timeout` and describe what the variables are for. I do not fully understand the purpose of `$user_selected`

File: tests/x11/multi_users_dm.pm [hunk 7]
```diff
@@ -38,11 +38,17 @@ sub ensure_graphical_target {
 }
 
 sub restart_x11 {
-    ensure_multi_user_target;
-    ensure_graphical_target;
+    my ($self, $setnologin) = @_;
+    record_soft_failure('Workaround for graphical errors on cirrus when restarting X11 -> performing reboot (see bsc#1063060)');
```
Comment (okurz) [inline tests/x11/multi_users_dm.pm]: Please put the bug id first, e.g. 'bsc#1234 Workaround for graphical errors…'

File: tests/x11/multi_users_dm.pm [hunk 8]
```diff
@@ -38,11 +38,17 @@ sub ensure_graphical_target {
...
+    type_string "reboot\n";
+    my $timeout = get_var('UEFI') ? 140 : 80;
```
Comment (okurz) [inline tests/x11/multi_users_dm.pm]: Why should restart_x11 define a different timeout for UEFI? If UEFI should really differ although I do not know why I suggest to put it into `wait_boot` itself. At best same for `handle_login`

File: tests/x11/multi_users_dm.pm [hunk 9]
```diff
@@ -38,11 +38,17 @@ sub ensure_graphical_target {
...
+    $self->wait_boot(bootloader_time => $timeout, nologin => $setnologin, reallynologin => $setnologin);
+    #ensure_multi_user_target;
+    #ensure_graphical_target;
```
Comment (okurz) [inline tests/x11/multi_users_dm.pm]: Don't add disabled code as nobody but you will understand why this is not active. Better leave it out and keep in your local git repository, either in `git stash` or a temporary "WIP"-commit

File: tests/x11/multi_users_dm.pm [hunk 10]
```diff
@@ -38,11 +38,17 @@ sub ensure_graphical_target {
 }
 
 sub restart_x11 {
-    ensure_multi_user_target;
-    ensure_graphical_target;
+    my ($self, $setnologin) = @_;
```
Comment (okurz) [inline tests/x11/multi_users_dm.pm]: How about:

```perl
my ($self, %args) = @_;
$args{setnologin} //= 0;
```

and use like `$self->restart_x11(setnologin => 1)`

File: lib/opensusebasetest.pm [hunk 11]
```diff
@@ -311,6 +311,7 @@ sub wait_boot {
     my $ready_time      = $args{ready_time} // 200;
     my $in_grub         = $args{in_grub} // 0;
     my $nologin         = $args{nologin};
+    my $reallynologin   = $args{reallynologin};
```
Comment (okurz) [inline lib/opensusebasetest.pm]: Is this a joke? I don't get it :D Please find a better name and describe it in the above documentation string

**Comment by asdil12** [comment]
Verification run: http://artemis.suse.de/tests/204

**Comment by asdil12** [comment]
ok to merge?

**Pull request merged**